File: Sources/Respring/VectorArithmetic/VectorArithmetic.swift
//
//  Created by ktiays on 2024/11/21.
//  Copyright (c) 2024 ktiays. All rights reserved.
// 

import Foundation

/// A type that can serve as the animatable data of an animatable type.
///
/// VectorArithmetic extends the AdditiveArithmetic protocol with scalar multiplication
/// and a way to query the vector magnitude of the value.
public protocol VectorArithmetic: AdditiveArithmetic {
    
    /// Returns the dot-product of this vector arithmetic instance with itself.
    var magnitudeSquared: Double { get }
    
    /// Multiplies each component of this value by the given value.
    mutating func scale(by rhs: Double)
}

extension VectorArithmetic {
    
    /// Returns a value with each component of this value multiplied by the given value.
    public func scaled(by rhs: Double) -> Self {
        var result = self
        result.scale(by: rhs)
        return result
    }
    
    /// Interpolates this value with `other` by the specified amount.
    ///
    /// This is equivalent to `self = self + (other - self) * amount`.
    @inlinable
    public mutating func interpolate(towards other: Self, amount: Double) {
        self = self + (other - self).scaled(by: amount)
    }
    
    /// Returns this value interpolated with `other` by the specified amount.
    ///
    /// This result is equivalent to `self + (other - self) * amount`.
    @inlinable
    public func interpolated(towards other: Self, amount: Double) -> Self {
        self + (other - self).scaled(by: amount)
    }
}

File: src/vector_arithmetic.rs
use crate::additive_arithmetic::AdditiveArithmetic;

/// A type that can serve as the animatable data of an animatable type.
pub trait VectorArithmetic: AdditiveArithmetic + Clone {
    /// Returns the dot-product of this vector arithmetic instance with itself.
    fn magnitude_squared(&self) -> f64;

    /// Multiplies each component of this value by the given value.
    fn scale_by(&mut self, scalar: f64);

    /// Returns a value with each component of this value multiplied by the given value.
    #[inline]
    fn scaled_by(mut self, scalar: f64) -> Self {
        self.scale_by(scalar);
        self
    }
}

macro_rules! vector_arithmetic_impl {
    ($($t:ty)*) => ($(
        impl VectorArithmetic for $t {
            fn magnitude_squared(&self) -> f64 {
                (*self as f64) * (*self as f64)
            }

            fn scale_by(&mut self, scalar: f64) {
                *self = (*self as f64 * scalar) as Self;
            }
        }
    )*)
}

vector_arithmetic_impl! { f32 f64 }
